harness/determined/common/experimental/experiment.py

```diff
@@ -115,6 +119,49 @@ def reload(self) -> None:
         resp = bindings.get_GetExperiment(self._session, experimentId=self.id).experiment
         self._hydrate(resp)
 
+    def add_label(self, label: str) -> None:
+        """Add a label to the experiment.
+
+        Arguments:
+            label: a string label to add to the experiment. If the label already exists,
+                the method call will be a no-op.
+        """
+
+        # URL-encode label for request.
+        label = parse.quote(label)
+        resp = bindings.put_PutExperimentLabel(
+            session=self._session, experimentId=self.id, label=label
+        )
+        self.labels = list(resp.labels)
```

Comment:
Which opinion are you looking for?
- I think that the server is just fundamentally broken because `set` mechanics cannot be safely used to implement `add/delete` mechanics, but add/delete mechanics are the only mechanics that we want to offer to end users.  So just fix the server.
- If the server were fixed, who cares if you replace the list or append to the list?  Maybe it matters because it's ordered?  Maybe a set would make more sense here, since that's really the experience we are trying to offer.
- In a world where the server isn't fixed, replacing self.labels with the freshest known value seems best, even if add_label() caused unexpected results to the ExperimentReference.labels value.  I'd rather be shown how other clients were breaking my logic then to be told things worked as I expected when really they didn't.

Reply:
Thanks, Ryan!

We've got PATCHes, PUTs, and POSTs all over our API with "set" semantics that serve as our standard way of adding/removing from collections. Until this comment, I don't think it'd occurred to me that this could be fixed. Even though that's exactly what Anda did right here.

I get a little confused with the "fix" language re: the server. At the moment, in this PR both `add_label` and `remove_label` both use add/remove methods on the backend. Even without "set", the local client's version can fall out of sync with the source of truth. I haven't figured out relevance of the antecedent in "In a world where the server isn't fixed, replacing self.labels with the freshest known value seems best."

Which is the precise question here I meant to pose:
1. Refresh the whole local snapshot of the collection?
2. Just add/remove the requested item?
3. Something else entirely?

I think you're right in weighing 1 vs. 2 to say that the user gets unhappy and surprised either way, and better to make that visible. I guess if we wanted we could also warn when the response indicates there's been a data update this client didn't know about.